### PR TITLE
gh-84978: Add float.from_number() and complex.from_number()

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
They are alternate constructors which only accept numbers
(including objects with special methods `__float__`, `__complex__`
and `__index__`), but not strings.

https://bugs.python.org/issue40801

* Issue: gh-84978

### CODE REVIEW

**Comment by mdickinson** [comment]
@serhiy-storchaka Are you still interested in pursuing this? Looks like it's still a draft; what's needed to make it ready for review?

**Comment by serhiy-storchaka** [comment]
It needs documentation. I'll write it if the rest be set.

It differs from your original idea (class methods instead of `operator` functions), so I do not know what you think about this. The spelling is also can be discussed (`from_number` vs `fromnumber`, or maybe something other?) Should `int.from_number` be added?

Should we add also special constructors for parsing? `int.parse` or `int.from_string`, etc. It can be done in a different PR, but it is useful to keep it in mind.

**Comment by serhiy-storchaka** [comment]
This tests refactoring in test_float/complex is required for testing `from_number()` methods. Without it they are irrelevant. Why do you need this?

**Comment by skirpichev** [comment]
> Why do you need this?

I was thinking about similar refactoring while improving test coverage for the complexobject.c.  So, maybe this could be useful regardless of the fate of this pr.

**Comment by serhiy-storchaka** [comment]
@mdickinson I added docs and remove the draft status. Hope it will help with review.

File: Doc/library/functions.rst [hunk 1]
```diff
@@ -387,6 +387,8 @@ are always available.  They are listed here in alphabetical order.
    to :meth:`~object.__float__`.  If :meth:`!__float__` is not defined then it falls back
    to :meth:`~object.__index__`.
 
+   See also :meth:`complex.from_number` which only accept single numeric argument.
```
Comment (mdickinson) [inline Doc/library/functions.rst]: Grammar nitpick:
```suggestion
   See also :meth:`complex.from_number` which only accepts a single numeric argument.
```

File: Doc/library/functions.rst [hunk 2]
```diff
@@ -684,6 +686,8 @@ are always available.  They are listed here in alphabetical order.
    ``x.__float__()``.  If :meth:`~object.__float__` is not defined then it falls back
    to :meth:`~object.__index__`.
 
+   See also :meth:`float.from_number` which only accept numeric argument.
```
Comment (mdickinson) [inline Doc/library/functions.rst]:
```suggestion
   See also :meth:`float.from_number` which only accepts a numeric argument.
```

File: Doc/library/stdtypes.rst [hunk 3]
```diff
@@ -624,6 +624,23 @@ Additional Methods on Float
 The float type implements the :class:`numbers.Real` :term:`abstract base
 class`. float also has the following additional methods.
 
+.. classmethod:: float.from_number(x)
+
+   Class method to return a floating point number constructed from a number *x*.
```
Comment (mdickinson) [inline Doc/library/stdtypes.rst:630]: Super-nitpick: I prefer the spelling `floating-point number`, with a hyphen (the "floating-point" part acts as a compound adjective). I think that's what we _mostly_ use throughout the docs (though no doubt there are exceptions).

File: Doc/whatsnew/3.13.rst [hunk 4]
```diff
@@ -75,6 +75,11 @@ New Features
 Other Language Changes
 ======================
 
+* Added class methods :meth:`float.from_number` and :meth:`complex.from_number`
+  to convert a number to :class:`float` or :class:`complex` type correspondingly.
+  They raise error if the argument is a string.
```
Comment (mdickinson) [inline Doc/whatsnew/3.13.rst]:
```suggestion
  They raise an error if the argument is a string.
```

**Review by mdickinson — APPROVED** [review]
This LGTM, and works as expected in my manual testing. I only have a couple of grammar nitpicks.

Thank you!

File: Doc/library/stdtypes.rst (hunk 3 above)
Comment (serhiy-storchaka) [reply Doc/library/stdtypes.rst:630]: This descriptions was simply copied from the `float` constructor description.

"floating-point" occurs 118 time, "floating point" occurs at least 159 times (in several cases it is split between lines, so it is not easy to get accurate number).

So for now I leave "floating point" for consistency with the constructor description. We will solve this in a separate issue.

**Pull request merged**